### DISCUSSION TITLE
Can I format (non-block) raw text with a colored background and with line wrapping?

### DISCUSSION THREAD

**Discussion by deffi** [typst/typst#2911] [Q&A] [answered]
### Goal

[Here](https://gist.github.com/deffi/6d5330c9e29d38d44c5b6e5befba8379)'s how GitHub renders inline code (in a Markdown document):
![grafik](https://github.com/typst/typst/assets/470243/5765913a-3bd2-4e3c-966b-7e780ce79854)

Note that:

- Code is rendered on a light grey background
- The background has rounded corners
- The code is line-wrapped
- Where the code is line-wrapped, the corners are not rounded

I want to achieve the same thing (plus syntax highlighting) with Typst, but my attempts were not successful.

The code examples below are also in [this document](https://typst.app/project/rtxQd1WEuvrjB0gmgFro9Q) in the Typst webapp.

### Baseline

````typst
Short code: `lorem` inside a paragraph

Longer code: ```python while Lorem in ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt``` inside a paragraph
````

Result: 
![grafik](https://github.com/typst/typst/assets/470243/21bf67f4-3523-473b-8140-39ff4aaeb368)

Line-wrapping is OK, but of course, there is no grey background

### Block

````typst
#show raw.where(): it => block(
  fill: rgb("ddd"),
  inset: 8pt,
  radius: 5pt,
  it
)

Short code: `lorem ipsum` inside a paragraph

Longer code: ```python while Lorem in ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt``` inside a paragraph
````

Result:
![grafik](https://github.com/typst/typst/assets/470243/2aec7976-bcf1-422f-9c3f-8c2b56553d8b)

The background looks good, but the raw text is in a separate paragraph.

### Box

````typst
#show raw: it => box(
  fill: rgb("ddd"),
  outset: 4pt,
  radius: 5pt,
  it
)

Short code: `lorem ipsum` inside a paragraph

Longer code: ```python while Lorem in ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt``` inside a paragraph
````

Result:
![grafik](https://github.com/typst/typst/assets/470243/76d78621-4705-49ff-bcba-5934a42c3ebf)

Raw text that does not fit on a line introduces line breaks before and after the raw text.

### Highlight

````typst
#show raw.where(): it => highlight(
  fill: rgb("ddd"),
  extent: 4pt,
  it
)

Short code: `lorem ipsum` inside a paragraph

Longer code: ```python while Lorem in ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt``` inside a paragraph
````

Result:
![grafik](https://github.com/typst/typst/assets/470243/45ef30b0-1186-47a4-be61-e91251335d5e)

This...
- ...does not support rounded boxes
- ...does not support vertical adjustment (I want the background to extend slightly above and below the text)
- ...fails with syntax highlighting
- ...extends into the regular text horizontally rather than creating space

**Comment by yochem** [comment]
See https://github.com/typst/typst/discussions/845. It seems not really possible currently, but you can hack it together by boxing every character or word.

**Comment by xkevio** [comment, marked as answer]
@yochem has the right idea! There might be issues with multiple spaces but you can adjust that, try this:

```typst
#show raw.where(block: false): r => {
  let words = r.text.split(" ")
  for (idx, word) in words.enumerate() {
    let w-radius = if words.len() == 1 {
      3pt
    } else if idx == 0 {
      (left: 3pt)
    } else if idx == words.len() - 1 {
      (right: 3pt)
    } else {
      0pt
    }
    
    box(
      fill: luma(245),
      outset: (y: 3pt),
      inset: (x: 2pt),
      radius: w-radius,
      raw(lang: r.lang, word)
    )
  }
}


Short code: `lorem ipsum` inside a paragraph

Longer code: ```python while Lorem in ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt``` inside a paragraph
```

**Reply by yochem** [reply]
Nice solution! 

It would be nice though if `higlight` supported the same parameters as box for the radius and outset/inset :)

**Reply by laurmaedje** [reply]
Nice idea!

**Reply by laurmaedje** [reply]
You can open a feature request for that if you want.

**Reply by yochem** [reply]
#2934

**Reply by deffi** [reply]
Also, #2939

**Reply by deffi** [reply]
This no longer works with Typst 0.11:

```
error: maximum show rule depth exceeded
   ┌─ raw-test.typ:19:6
   │
19 │       raw(lang: r.lang, word)
   │       ^^^^^^^^^^^^^^^^^^^^^^^
   │
   = hint: check whether the show rule matches its own output
```

Given that we have a show rule for raw elements that generates more raw elements, this isn't even surprising. But why did it work in Typst 0.10?

Can we indicate that the same show rule should not be applied to the generated elements?

**Reply by xkevio** [reply]
It only worked before because the `block` parameter wasn't synthesized, I think. Meaning, that `raw` and `raw(block: false)` were "different" things to the compiler even though `block: false` is the default. So mostly a coincidence

**Reply by th1j5** [reply]
For others:
Since https://github.com/typst/typst/pull/3526 is merged, highlight is the best candidate.
There is still a bug mentioned in this pull request about highlighting done by word, instead of continuously.
If you set the extent low enough and the radius too, you can already have a decent result.
```typ
#highlight(radius: 1pt, extent: 0.5pt, fill:silver, top-edge: 1.1em, bottom-edge: -0.3em)[```python while true: pass; ABC```]
```
![image](https://github.com/user-attachments/assets/c26cf678-d40d-4d8c-8f5a-9ae473fc36c7)
- [x] Rounded boxes
  - [ ] Also on line ends... actually, due to the mentioned bug, at each word boundary and each space. Not visible due to low radius and extent > 0.
- [x] Vertical adjustment
- [x] Has syntax highlighting, I'm not sure what you ever meant with this... If you're talking about the background covering the syntax-highlighted words, this is solved by choosing small radius & extent (<1pt) values, until that bug is fixed.
- [x] You can create space manually around a highlighted code block.

Edit: after some experimentation, that bug is really annoying, so the mentioned settings are actually the only settings giving a good result for the highlighting.
Bug: https://github.com/typst/typst/issues/2939
Synthesis bug report about highlighting: https://github.com/typst/typst/issues/5872

**Comment by ArcusCerebellumus** [comment]
There's one more option if you're fine with using nerd-fonts (and don't need fancy styles with outlines within the code block itself), that worked for my case: 
```typ
#show raw: content => text(
    font: "CaskaydiaCove NF",
    size: 10pt
)[
    #let f = rgb("#fafafa")
    // here I adjusted size and baseline of "caps" to fit my particular font
    // as much as I considered enough for them to not stick out like sore thumb
    #let cap(c) = text(baseline: -0.04pt, size: 9.91pt, fill: f)[#c]
    #cap[#str("\u{E0B6}")]#highlight( fill: f)[#content]#cap[#str("\u{E0B4}")]
]
```
results in something like this:

![Screenshot_1523](https://github.com/user-attachments/assets/0166e147-4a25-45f2-b32e-c123b0791e82)

**Comment by dkchshv** [comment]
Here's another try. But it's unbreakable.
```typst
#show raw.where(block: false): string => {
  let x_margin_inner = 3pt
  let x_margin_outer = 5pt
  let Ymargin = 2pt
  let increaseLineHeight = false
  
  box(
    inset: (x: x_margin_outer, y: Ymargin * int(increaseLineHeight)),
    baseline: Ymargin * int(increaseLineHeight),
    box(
      width: measure(string).width + 0pt,
      outset: (x: x_margin_inner, y: Ymargin),
      radius: 15%,
      fill: gray.lighten(60%),
      stroke: .7pt+gray,
      align(center, string)
    )
  )
}
```
<img width="1713" height="794" alt="изображение" src="https://github.com/user-attachments/assets/726e59b3-d8b4-41f7-9647-d4a2d2c19bc7" />
